cmd/site-replication.go

```diff
@@ -252,73 +252,122 @@ const (
-				return madmin.ReplicateAddStatus{}, errSRBackendIssue(err)
+				return psi, errSRBackendIssue(err)
 			}
 			if len(res) > 0 {
-				localHasBuckets = true
+				pi.Empty = false
+			}
+			pi.self = true
+		} else {
+			s3Client, err := getS3Client(v)
+			if err != nil {
+				return psi, errSRPeerResp(fmt.Errorf("unable to create s3 client for %s: %w", v.Name, err))
+			}
+			buckets, err := s3Client.ListBuckets(ctx)
+			if err != nil {
+				return psi, errSRPeerResp(fmt.Errorf("unable to list buckets for %s: %v", v.Name, err))
+			}
+
+			if len(buckets) > 0 {
+				pi.Empty = false
 			}
-			continue
 		}
+		psi = append(psi, pi)
+	}
+	return
+}
 
-		s3Client, err := getS3Client(v)
+// AddPeerClusters - add cluster sites for replication configuration.
+func (c *SiteReplicationSys) AddPeerClusters(ctx context.Context, psites []madmin.PeerSite) (madmin.ReplicateAddStatus, SRError) {
+	sites, serr := c.getSiteStatuses(ctx, psites)
+	if serr.Cause != nil {
+		return madmin.ReplicateAddStatus{}, serr
+	}
+	var (
+		currSites            madmin.SiteReplicationInfo
+		currDeploymentIDsSet = set.NewStringSet()
+		err                  error
+	)
+	if c.enabled {
+		currSites, err = c.GetClusterInfo(ctx)
 		if err != nil {
-			return madmin.ReplicateAddStatus{}, errSRPeerResp(fmt.Errorf("unable to create s3 client for %s: %w", v.Name, err))
+			return madmin.ReplicateAddStatus{}, errSRBackendIssue(err)
 		}
-		buckets, err := s3Client.ListBuckets(ctx)
-		if err != nil {
-			return madmin.ReplicateAddStatus{}, errSRPeerResp(fmt.Errorf("unable to list buckets for %s: %v", v.Name, err))
+		for _, v := range currSites.Sites {
+			currDeploymentIDsSet.Add(v.DeploymentID)
+		}
+	}
+	deploymentIDsSet := set.NewStringSet()
+	localHasBuckets := false
+	nonLocalPeerWithBuckets := ""
+	var selfIdx = -1
+	for i, v := range sites {
+		// deploymentIDs must be unique
+		if deploymentIDsSet.Contains(v.DeploymentID) {
+			return madmin.ReplicateAddStatus{}, errSRInvalidRequest(errSRDuplicateSites)
 		}
+		deploymentIDsSet.Add(v.DeploymentID)
 
-		if len(buckets) > 0 {
+		if v.self {
+			selfIdx = i
+			localHasBuckets = !v.Empty
+			continue
+		}
+		if !v.Empty && !currDeploymentIDsSet.Contains(v.DeploymentID) {
 			nonLocalPeerWithBuckets = v.Name
 		}
 	}
-
+	if c.enabled {
+		// If current cluster is already SR enabled and no new site being added ,fail.
+		if currDeploymentIDsSet.Equals(deploymentIDsSet) {
+			return madmin.ReplicateAddStatus{}, errSRInvalidRequest(errSRCannotJoin)
+		}
+		if len(currDeploymentIDsSet.Intersection(deploymentIDsSet)) != len(currDeploymentIDsSet) {
+			diff := currDeploymentIDsSet.Difference(deploymentIDsSet)
+			var diffSlc []string
+			for _, v := range currSites.Sites {
+				if diff.Contains(v.DeploymentID) {
+					diffSlc = append(diffSlc, v.Name)
+				}
+			}
+			return madmin.ReplicateAddStatus{}, errSRInvalidRequest(fmt.Errorf("All cluster sites must be specified - missing %s", strings.Join(diffSlc, " ")))
+		}
+	}
```

Comment:
We perhaps need some unit tests for these parts - having a separate function would help.

Reply:
refactored and added tests